### ISSUE TITLE
frames do not match in 'not top level in wrap-native for #<subr read-event>' (root cause: flyspell runs `sit-for` in `post-command-hook`)

### ISSUE THREAD

**Issue by lastquestion** [lastquestion/explain-pause-mode#54]
Reported by @condy0919 in #50, splitting it into a separate bug:
```
Explain-pause-mode: please report this bug by creating a Github
issue at https://github.com/lastquestion/explain-pause-mode. Explain-pause-mode
is now _disabled_ so you can continue to hopefully use Emacs. Info follows:


frames do not match in 'not top level in wrap-native for #<subr read-event>'
current:
#s(explain-pause-command-record root-emacs nil nil 1146334 (24320 27529 114676 750000) nil nil nil nil 0)
test:
#s(explain-pause-command-record root-emacs nil nil 1146334 (24320 27529 114676 750000) nil nil nil nil 0)


Backtrace:
  backtrace()
  explain-pause-report-measuring-bug("not top level in wrap-native for #<subr read-event..." #s(explain-pause-command-record :command root-emacs :native nil :parent nil :executing-time 1146334 :entry-snap (24320 27529 114676 750000) :too-slow nil :is-profiled nil :under-profile nil :profile nil :depth 0) #s(explain-pause-command-record :command root-emacs :native nil :parent nil :executing-time 1146334 :entry-snap (24320 27529 114676 750000) :too-slow nil :is-profiled nil :under-profile nil :profile nil :depth 0))
  explain-pause--wrap-native(#<subr read-event> nil t 3)
  apply(explain-pause--wrap-native #<subr read-event> (nil t 3))
  read-event(nil t 3)
  sit-for(3)
  flyspell-check-word-p()
  flyspell-post-command-hook()
```

**Comment by lastquestion** [comment]
Post-command-hook called from C code in `keyboard.c` at https://github.com/emacs-mirror/emacs/blob/master/src/keyboard.c#L1257, https://github.com/emacs-mirror/emacs/blob/master/src/keyboard.c#L1477. These are outside `call-interactively` which comes from `command_execute` at https://github.com/emacs-mirror/emacs/blob/master/src/keyboard.c#L1463

All `safe_run_hooks` in `emacs src`:
```
w32menu.c:2
xdisp.c:2
minibuf.c:1
fileio.c:1
keyboard.c:17
alloc.c:1
xmenu.c:2
lisp.h:1
emacs.c:1
window.c:1
```

not all of them are dangerous. Unique hooks called:
|Hook|Outside of command loop|Checked|Fixed|
|---------|----------------------------|----------|---------|
|Qactivate_menubar_hook||||
|Qmenu_bar_update_hook||||
|Qminibuffer_exit_hook||||
|fileio.c: hook||||
|Qpost_command_hook||||
|Qdelayed_warnings_hook||||
|Qdeferred_action_function||||
|Qecho_area_clear_hook||||
|Qpre_command_hook||||
|Qpost_command_hook||||
|safe_run_hooks_error (keyboard.c)||||
|Qecho_area_clear_hook||||
|Qpost_gc_hook||||
|Qkill_emacs_hook||||
|Qwindow_state_change_hook||||

I'm pretty sure the only real dangerous ones are in `keyboard.c`, so going to focus on these in an immediate PR. Most other hooks are run as a result of something that will already pass through `call-interactively`.

**Comment by lastquestion** [comment]
It's possible someone can just modify the hook list directly without calling `add-hook`. No one ought to be such a bad elisp citizen, but they might... as a backup, can we use the same technique as the debugger and break on changes to core hook lists called from C code?

-> add-hook catches 99%
-> evil people get caught for calls from native code. we don't get any more crashes 
-> some evil people will hide their execution cost in other people's calls if they directly add to their lists, but what can we do...

**Comment by lastquestion** [comment]
Let's focus this issue on hooks run from native code only. We have two `run_hook_with_args` family and `safe_run_hooks` family. 

* `safe_run_hooks` has a list that isn't that long, and most of them shouldn't even be a problem. 
- [ ] Make a list of dangerous ones
- [ ] Advise `add-hook`; when the hook is "dangerous", advise that hook to wrap itself in a new frame; 
- [ ] On enable, run through all the "dangerous hook lists", advising them.
* `run_hook_with_args` family mostly calls elisp so we can advise `run-hooks` at some level in elisp land. Deal with that in a separate issue.

**Comment by lastquestion** [comment]
So, this works with `pre-command-hook` and `post-command-hook` and `flyspell` now works (I tested manually).

I will push a PR to fix this to unblock people and then get a test suite added and then verify all the other 12 places are or are not dangerous.

**Comment by lastquestion** [comment]
Event stream when working:
```
("enter" "flyspell-post-command-hook" "root-emacs" "post-command-hook" nil 0 nil nil nil 2)
("enter" "#<subr read-event>" "flyspell-post-command-hook" "flyspell-post-command-hook" t 0 nil nil nil 3)
("exit" "#<subr read-event>" "flyspell-post-command-hook" 1324 nil)
("exit" "flyspell-post-command-hook" "post-command-hook" 0 nil)
```

Excitingly, this also means `post-command-hook` actors show up in the `top` buffer:
```
flyspell-post-command-hook                                                                                                                                         0   0.52 32    61
```

**Comment by lastquestion** [comment]
* `Qactivate_menubar_hook` | `Qmenu_bar_update_hook` -> startup / server / redisplay. ❓ 
* `Qminibuffer_exit_hook` -> `read_minibuf` -> already covered ✔️ 
* `auto-save-hook` -> `do-auto-save` and `shut_down_emacs` ❓ 
* `Qdelayed_warnings_hook`, `Qdeferred_action_function`, `Qecho_area_clear_hook` -> command-loop ❌  needs hooking
* `Qpost_gc_hook` -> from GC -> from alloc.c ❓ 
* `Qkill_emacs_hook` -> `kill-emacs` | `Fkill_emacs` ❓ 
* `Qwindow_state_change_hook` -> redisplay ❓ 

❓ = if some idiot ran a read-key code in these functions, they well deserve to be alerted
✔️ = already covered, but perhaps we should hook it anyway - I think people run lots of code here
❌ = not covered, needs covering

**Comment by lastquestion** [comment]
`delayed-warnings-hook` claims `;; At present this is used only for Emacs internals.`, but it's documented and used by Doom at least.
`deferred-action-function` - deprecated, but is replaced by `post-command-hook`
`echo-area-clear-hook` - seems rarely used

Let's hook all 3 anyway.

**Comment by lastquestion** [comment]
Adding `post_gc_hook` to the list, I can imagine someone printing a message and waiting.

**Comment by lastquestion** [comment]
@condy0919 this should be fixed now 💯

**Comment by condy0919** [comment]
Thanks for your work!
